docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,77 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability of extending the DNS configuration of the gateway with the DNSPolicy resource. This resource allows users to configure health checks. As a result of configuring health checks, the controller creates the health checks in Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it becomes unhealthy, and enabling it again when it becomes healthy again.
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* ~~Transparently keep support for other health check providers like Route53~~
+
+## Proposal
+
+#### `DNSPolicy` resource
+
+The presence of the `healthCheck` field in the DNSPolicy will affect the reconciliation
+of the endpoints when creating/updating DNSRecords:
+
+* For every endpoint that is generated, a health check is created based on the configuration in the DNSPolicy
+* Relevant `providerSpecific` fields are added to the endpoint:
+  * A reference to the equivalent `DNSHealthCheckProbe` in the downstream cluster
+  will be set: `health-probe/name`, `health-probe/namespace`
+  * The health status of the `DNSHealthCheckProbe` will be propagated
+
+A `failureThreshold` field will be added to the health spec, allowing users
```

Comment:
@philbrookes I wonder should we have a successThreshold? IE when do we consider an endpoint healthy again? one 200 response is likely not enough to assume it is back to ready??